### PR TITLE
feat: implement multiline support (continued)

### PR DESCRIPTION
**Pull request by joshuakb2**
I saw #51 and thought it was a good idea, so I thought I'd try to get it over the finish line.

I rebased aurelien-brabant's commits onto main as they had fallen behind. I fixed a typo in his code and changed the `RHS` index to use `std::string::at` like you suggested.

> what if `line.find_last_not_of(MULTILINE_SPACE_CHARSET)` returns `std::string::npos`

This appears to be a non-issue. In that case the code is evaluated as `line.substr(0, 0)`.

I also added another multiline test for configs that contain a whole empty line in the middle, and I fixed a bug where the `RHS` was assumed to be non-empty.

Let me know if you still have concerns about this feature and I'll see if I can address them.

### CODE REVIEW

**Review by vaxerski — COMMENTED** [review]
considering multi-line stuff can _only_ happen when parsing a config file, I am in favor of moving them out of the ::parseLine func altogether. This adds needless complexity.

I would instead move it to parseFile and parseRawStream

**Comment by joshuakb2** [comment]
In that case, I'll take another look at this change and see about implementing it the way you're describing. Thanks for the quick feedback!

**Comment by joshuakb2** [comment]
I implemented a simpler version of multi-line support in `parseFile` and `parseRawStream`.

Functional changes from the previous implementation:

- The `\` at the end of the line may not be followed by a comment or whitespace.
- Lines ending with `>` are not treated specially at all.

I should also point out that this feature is a breaking change because people may have config lines ending with backslashes where those backslashes are currently treated as normal text. But this was true of the previous implementation as well.

I welcome any feedback, and thanks again for considering this pull request!

File: src/config.cpp [hunk 1]
```diff
@@ -695,22 +695,52 @@ CParseResult CConfig::parse() {
...
-    int               linenum = 1;
+    std::string rawLine    = "";
+    int         rawLineNum = 0;
+    std::string line       = "";
+    int         lineNum    = 0;
 
     std::stringstream str(stream);
 
-    while (std::getline(str, line)) {
+    while (std::getline(str, rawLine)) {
```
Comment (vaxerski) [inline src/config.cpp:705]: instead of getline into rawLine and then copying the string into line, why not just getline to line? :)

File: src/config.cpp [hunk 2]
```diff
@@ -736,21 +766,50 @@ CParseResult CConfig::parseFile(const char* file) {
...
+                gotWholeLine = false;
+                break;
+            }
+
+            ++rawLineNum;
+            line += rawLine;
+            mergeWithNextLine = line.length() > 0 && line.at(line.length() - 1) == '\\';
+        }
+
+        if (!gotWholeLine)
```
Comment (vaxerski) [inline src/config.cpp:799]: considering all of these are duplicated, can we make this into a helper function that takes a ref to iffile and returns an std::expected<std::string, int>? If it returns an int, it means fail (break and die)

Comment (joshuakb2) [reply src/config.cpp:799]: Sounds good. I actually tried to make a helper function at first, but I consolidated too much logic and accidentally changed the behavior of one of the functions. I'll take another stab at it.

**Review by vaxerski — COMMENTED** [review]
rest lgtm

**Comment by joshuakb2** [comment]
Cool! I've addressed those changes you asked for. Thanks!

File: include/hyprlang.hpp [hunk 3]
```diff
@@ -446,6 +447,12 @@ namespace Hyprlang {
 
         CConfigImpl* impl;
 
+        enum eGetNextLineFailure : uint8_t {
+            GETNEXTLINEFAILURE_EOF = 0,
+            GETNEXTLINEFAILURE_BACKSLASH,
+        };
+
+        std::expected<std::string, eGetNextLineFailure> getNextLine(std::istream& str, int &rawLineNum, int &lineNum);
```
Comment (vaxerski) [inline include/hyprlang.hpp:455]: The enum and function don't belong in CConfig, IMO.

They don't use the locals at all, and only add a burden of `<expected>` on the TU.

I'd throw both of these into the cpp file only, as a local, `static` function

**Review by vaxerski — COMMENTED** [review]
rest lgtm

File: include/hyprlang.hpp (hunk 3 above)
Comment (joshuakb2) [reply include/hyprlang.hpp]: Makes sense to me! Done.

File: src/config.hpp [hunk 4]
```diff
@@ -65,6 +66,13 @@ struct SSpecialCategory {
     size_t anonymousID = 0;
 };
 
+enum eGetNextLineFailure : uint8_t {
+    GETNEXTLINEFAILURE_EOF = 0,
+    GETNEXTLINEFAILURE_BACKSLASH,
+};
+
+static std::expected<std::string, eGetNextLineFailure> getNextLine(std::istream& str, int &rawLineNum, int &lineNum);
```
Comment (vaxerski) [inline src/config.hpp:74]: this is not used outside of config.cpp, this can be removed altogether

**Comment by joshuakb2** [comment]
Alright, I've fixed that now. Thanks for the feedback, I hadn't considered that. I don't write much C++ these days but it's good to get a little practice in.

**Review by vaxerski — APPROVED** [review]
thanks!
